### ISSUE TITLE
Determine if `onData` argument name makes sense

### ISSUE THREAD

**Issue by jerelmiller** [apollographql/apollo-client#10142]
With the change in #10134, the function signature to `onData` changed slightly from the `onSubscriptionData` callback. `onSubscriptionData` took an `options` argument with a `subscriptionData` property that gave access to the subscription result. With the name change to `onData`, I felt it made sense to rename that property from `subscriptionData` to `data` to better align with the name of the callback. 

While the new property is more terse and better aligned with the name of the callback, this resulted in a small quirk. The `data` passed in is the full [`SubscriptionResult`](https://github.com/apollographql/apollo-client/blob/184ebc09966d657d2219f5212dc306fe246dfdaa/src/react/types/types.ts#L240-L247) type, which includes a `data` property of its own. Using this means accessing a double `data` property:

```ts
useSubscription(subscription, {
  onData: ({ data }) => {
    console.log(data.data);
  }
});
```

This also has the downside that using `data` means shadowing the `data` property returned from `useSubscription` (or `useQuery`, etc.)

```ts
const { data } = useSubscription(subscription, {
  onData: ({ data }) => {
    // `data` is shadowed from the outer `data` property
  }
});
```

There is an [ESLint rule](https://eslint.org/docs/latest/rules/no-shadow) that prohibits this behavior, which means aliasing the field when used:

```ts
const { data } = useSubscription(subscription, {
  onData: ({ data: subData }) => {
    // it is now ok to use `subData` as it no longer shadows the outer `data` variable
  }
});
```

We should determine if this makes sense or if we should consider renaming that `data` property to something else to avoid the double `data.data` access.

## Alternatives

**`subscriptionData`**

We could rename this back to `subscriptionData` to keep the callback signature the same as before. This at least gets rid of the double `data.data` access, but diverges from the name of the callback a bit.

**`result`**

We could also consider `result` as a possibility. This would better align with its type name [`SubscriptionResult`](https://github.com/apollographql/apollo-client/blob/release-3.7/src/react/types/types.ts#L207). The downside here is that this property now diverges a tad from `useQuery` which uses the [`data` name](https://github.com/apollographql/apollo-client/blob/release-3.7/src/react/types/types.ts#L43) in its callbacks.

**Comment by bignimbus** [comment]
I have a not-strong-at-all opinion that the duplication is acceptable here.  Are there any descriptive alternatives we could consider?

**Comment by jerelmiller** [comment]
@bignimbus it was named `subscriptionData` before, so one option could be to rename it back to the original. It still has the double `data` in the name, but not as bad as this. Another alternative would be to call this `result` which would better align with the [type](https://github.com/apollographql/apollo-client/blob/release-3.7/src/react/types/types.ts#L212) of this property. 

Of those two alternatives, I like `result`, the most, but it does diverge a bit from [`useQuery`](https://github.com/apollographql/apollo-client/blob/release-3.7/src/react/types/types.ts#L43) which uses the name `data` in its callbacks. 

I don't feel too strongly one way or another. While this is a small quirk, its not the end of the world.

**Comment by efstathiosntonas** [comment]
If we have a `const {data} =useQuery` in the same component then we have a shadow variable, I’ve just upgraded to 3.7 and used onData with `{data: subData}` all over the place.

**Comment by jerelmiller** [comment]
@efstathiosntonas that's a really good point! I've updated the issue above to note that.

**Comment by jerelmiller** [comment]
Seeing as this has been released into v3.7, any change now would constitute a breaking change. I'm going to close this issue and we can re-address once we are ready for v4 development.
